056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+and [PR #8892](https://github.com/strimzi/strimzi-kafka-operator/pull/8892).
+
+
+## Current situation
+
+When operating on a ZooKeeper based cluster the KafkaRoller has the behaviour 
+described below.
+
+### Order to roll nodes
+Roll in order:
+ 1. Unready brokers
+ 2. Ready brokers
+ 3. Current controller
+
+### Triggers
+The following are some of the triggers that roll a broker:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready
+- Broker's configuration has changed
+
+### Rollability checks
+ 1. Does not force roll a broker performing log recovery
+ 2. Attempts to connect an admin client to a broker and if it can't connect rolls the broker
+ 3. Force rolls a broker if the pod is in stuck state
+ 4. Rolls a broker if the configuration has changed and cannot be updated dynamically
+ 5. Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+
+#### Unready pod
+If pod is unready, KafkaRoller awaits its readiness before doing anything. If it hasn't become ready still, it checks whether it is stuck and considers to force roll. A pod is considered stuck if it is in one of following states:
+- `CrashLoopBackOff`
+- `ImagePullBackOff`
+- `ContainerCreating`
+- `Pending` and `Unschedulable`
+
+If pod is not stuck but unready, the next rollability checks are performed.
+
+
+#### Configuration changes
+ - Retrieves the current Kafka configurations of the broker via admin client and compares it with the desired configurations specified in the Kafka CR. 
+ - Performs dynamic configuration updates if possible
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+1. Unready controller/combined nodes
+2. Ready controller/combined nodes
+3. Active controller (applies to both pure controller and combined node)
+4. Unready broker-only nodes
+5. Ready broker-only nodes
```

Comment:
This order LGTM, but you've not stated the source of truth for knowing whether a particular pod is a pure controller node, combined node or pure broker node. The operator has its view (which is available from the `NodeRef`), and Kafka has its own view. We expect these to be consistent, but AFAIK that's not guaranteed (maybe @scholzj will correct me if there is such a guarantee).

Reply:
That is complicated. I guess the NodeRefs currently shows the desired state. However, right now it is the same as the current state given our lack of support for controller rolling etc. So there was no thought about this until now I guess.

Reply:
Actually, the Roller IIRC queries the pods to decide if they are ready or not. Maybe it can get the current state from their annotations that will have the current state until rolled.

Reply:
In terms of ordering I wonder if it is actually better to restart the pods based on the desired roles, not the current roles. For example, if a node was previously a `broker` and now is `controller,broker` you probably want to restart it before you begin restarting the other broker nodes so it can join the controller quorum. I think in each process role change restarting the node based on the desired roles is preferable.